keps/sig-cli/3895-kubectl-delete-interactivity/README.md

```diff
@@ -0,0 +1,926 @@
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+Introduce interactive mode for `kubectl delete` command, to allow users protect cluster
+administrators from irrevocable removal of critical resources.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+kubectl delete is disruptive and irreversible command and should be used cautiously. However, there
+are various circumstances such as inaccurate usage, mistyping, hasty tab completions, etc. that
+this command can position the cluster to a state where it is not intended to be. Due to the backwards
+compatibility concerns, seeking confirmation from a user before proceeding to genuinely delete as default, is out of the table.
+
+On the other hand, we need to provide a way to mitigate accidental deletes to users and this KEP
+proposes a new flag, namely interactive(-i) that will ask confirmation from a user to proceed with deletion.
+
+### Goals
+
+- Introduce interactive(-i) flag to be used for asking confirmation to user to proceed actual delete operation or not.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+- Smart solutions on client side to detect deletion is accidental or not
+- Deletion protection mechanism on server side
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+When a user passes `--interactive`/`-i` flag to `kubectl delete` command, all the resources planned to be deleted will be 
+shown to the user as preview. Command continues deletion process if user confirms by typing `y`. If user types `n` or
+any other value other than `y`, process will be cancelled and command returns a message with zero exit code.
+
+```sh
+$ kubectl delete --interactive -f test_deployment.yaml 
+You are about to delete the following 2 resource(s):
+Deployment/dockerd
+Deployment/dockerd2
+Do you want to continue? (y/n): n
+deletion is cancelled
+
+$ kubectl delete -i -f test_deployment.yaml
```

Comment:
Just curious, what was the reason behind passing the resource through a file. Could it be `namespace/resource_name` (for ns scoped) and `resource_name` (for cluster scoped)?

Also is the file name a required argument? Would `kubectl delete deployments --interactive` provide all the list of deployments in the cluster to be deleted?

Reply:
> Just curious, what was the reason behind passing the resource through a file. Could it be `namespace/resource_name` (for ns scoped) and `resource_name` (for cluster scoped)?
> 
It is a pretty standard way of passing resources to the commands and as you can see in https://github.com/kubernetes/kubernetes/blob/c5b95f60faa5dc2acfb5397e1a83909c24ec1791/staging/src/k8s.io/cli-runtime/pkg/genericclioptions/filename_flags.go#L62, it is added to all commands.

> Also is the file name a required argument? Would `kubectl delete deployments --interactive` provide all the list of deployments in the cluster to be deleted?

File name is not required and `kubectl delete deployment test --interactive` should also work.